keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,469 @@
+
+And a couple of hard-coded constants that specify how fast the cluster can scale up:
+
+- [scaleUpLimitFactor][] = 2.0
+- [scaleUpLimitMinimum][] = 4.0
+
+As a result, users can't influence scale velocity, and that is a problem for many people. There're several open issues in tracker about that:
+
+- [#39090][]
+- [#65097][]
+- [#69428][]
+
+[--horizontal-pod-autoscaler-downscale-stabilization-window]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+[scaleUpLimitFactor]: https://github.com/kubernetes/kubernetes/blob/release-1.14/pkg/controller/podautoscaler/horizontal.go#L55
+[scaleUpLimitMinimum]: https://github.com/kubernetes/kubernetes/blob/release-1.14/pkg/controller/podautoscaler/horizontal.go#L56
+[#39090]: https://github.com/kubernetes/kubernetes/issues/39090
+[#65097]: https://github.com/kubernetes/kubernetes/issues/65097
+[#69428]: https://github.com/kubernetes/kubernetes/issues/69428
+
+### Goals
+
+- Allow the user to be able to manage the scale velocity
+
+### Non-Goals
+
+TBA
+
+## Proposal
+
+We need to introduce an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behavior.
+For both direction (scale up and scale down) there should be a `Constraint` field with the following parameters:
+
+- `periodSeconds` - a parameter to specify the time period for the constraint, in seconds
+- `percent` - a parameter to specify the relative speed, in percentages:
+    i.e., if ScaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
+- `pods` - a parameter to specify the absolute speed, in the number of pods:
+    i.e., if ScaleUpPods = 5 , then we can add 5 more pods (10 -> 15)
+- `delay` - a parameter to specify the window over which the max (or min) recommendation is used. It behaves the same as the [Stabilization Window][].
+
+A user will specify the parameters for the HPA, thus controlling the HPA logic.
+
+If the user specifies two parameters, two constraints are calculated, and the largest is used (see the [The motivation to pick the largest constraint][] section below).
+
+If the user doesn't specify any parameter, the default value for that parameter is used (see the [Default Values][] section below).
+
+If the user set both `percent` and `pods` to `0`, it means that the corresponding resource (e.g. `deploy`, `rs`) should not be scaled in that direction.
+
+[The motivation to pick the largest constraint]: #the-motivation-to-pick-the-largest-constraint-concept
+[Default Values]: #default-values
+[Stabilization Window]: #stabilization-window
+
+### User Stories
+
+#### Story 1: Scale Up As Fast As Possible
+
+This mode is essential when you want to respond to a traffic increase quickly.
+
+Create an HPA with the following configuration:
+
+- `constraints`:
+  - `scaleUp`:
+    - `percent = 900`    (i.e., to increase the number of pods 10 times per minute is ok).
```

Comment:
For final API we may want to try a few variants to see what UX works best and is least confusing.  E.g. if I want to allow 10x growth, does it make sense to say "grow by 900%"  or "grow to 1000%" ?

For examples like "grow by 100%" it seems pretty obvious, just less so at larger numbers :)

Reply:
I would prefer "grow to 1000%".  It makes the math more obvious: `maxAllowed = current * (percent / 100)`